### DISCUSSION TITLE
nextjs first time load take while

### DISCUSSION THREAD

**Discussion by ahmed-el-amine** [vercel/next.js#34540] [Help] [answered]
in nextjs production when I run the server for the first time and create the first request 
[GET, POST,  ...], it take about 3-7 seconds to respond then the website run normally
i figured that next js create static generation on first user visit the website 
how to prevent this or how to trigger this event after run server so the first user no need to wait for this event to happen 

I have solve this by just create an custom express server and run axios request to home page to trigger this event after server starts

but in my case iam using **pm2** for production and wen i create more than 2 cluster the first cluster who trigger this event wen i call axios request for my domain and the other clusters still waiting for 3-7 sec at first visit

**Comment by 2stash** [comment]
@a7medmo7med9 From your description, my guess is that your production server goes to sleep if you don't have traffic. Which is why your custom express server works to speed up your page load. The resources serving your website go to sleep after not being used, but your express server wakes it up so that it can respond to incoming traffic immediately. 

Other problems could be how much data fetching you have in getStaticProps and getServerSideProps.

**Reply by ahmed-el-amine** [reply]
No my website doesn't sleep if there's no traffic

before i use custom server this still happens
Nextjs wait for the first user visit the website to start create static generation pages i think

after first request the server run normally until i restart it again

**Reply by 2stash** [reply]
@a7medmo7med9 Is below a correct summary of your issue?

- server is turned on
- first visitor takes 3-7 seconds
- second visitor is instant/fast
- third visitor is instant/fast
- server is restarted
- first visitor takes 3-7 seconds
- second visitor is instant/fast
- ...

Something like this?

**Reply by ahmed-el-amine** [reply]
exactly

**Reply by 2stash** [reply]
@a7medmo7med9 So I would say, if you found a simple solution then go with that.

But in terms of a next.js solution, it looks like first time loading does sometimes have problems depending on your application.

This issue appears to be related to your issue and has some suggestions and workarounds. My understanding is that the lazy loading react.js implements can sometimes backfire in certain use cases. If it is related add a comment in it!
https://github.com/vercel/next.js/issues/23187

Discussions similar to yours.
https://github.com/vercel/next.js/discussions/12447
https://github.com/vercel/next.js/discussions/29857

I would suggestion reading through these (they are not long) and see if any of the suggestions in there help you. I personally would try to find at what point, if any, the page loads quickly on first load. So if you put a brand new create-next-app on your server, does it still take 3-7 seconds to load? And as you build out your code at what point does the first load time go up. I understand this might not be feasible depending on the size of your app, but this line of thinking is how I would troubleshoot it.

Good luck!

**Reply by ahmed-el-amine** [reply]
Thank you for your help 

I just read about new version of next js v12.1.0 
And i update my project and this problem got solved 😃 i don't know how exactly happened in this new version but it working at first load not taking any time

**Reply by jodik** [reply]
2cents: Added 2 main pages of my project here: https://screenbud.com/shot/74e0dbaf-27b7-4e41-b4ff-0427c48b24bd initial loadup time went down from 10-20secs to ~0

**Comment by ahmed-el-amine** [comment, marked as answer]
I updated my app to use next js 12.1.0 and it fix my problem now first load not taking any time load fast as any other request

**Comment by mahmir217** [comment]
> in nextjs production when I run the server for the first time and create the first request [GET, POST, ...], it take about 3-7 seconds to respond then the website run normally i figured that next js create static generation on first user visit the website how to prevent this or how to trigger this event after run server so the first user no need to wait for this event to happen
> 
> I have solve this by just create an custom express server and run axios request to home page to trigger this event after server starts
> 
> but in my case iam using **pm2** for production and wen i create more than 2 cluster the first cluster who trigger this event wen i call axios request for my domain and the other clusters still waiting for 3-7 sec at first visit

What next version were you using earlier ?

**Reply by ahmed-el-amine** [reply]
i think it was v12.0.10 now im using 12.3.1

**Comment by kopach** [comment]
I'm using next.js `13.1.1` and having same issue.

**Comment by bensleym** [comment]
My first load time for SSR was 20s, it's taken over a week of digging and testing to finally get it below 1s to around 400-500ms
13.4.3 NextJs version

1) Added "use client" to client side components
2) Used Dynamic imports
4) modularizeImports for MUI (seemed to have the biggest impact)
modularizeImports: {
    "@mui/icons-material": {
      transform: "@mui/icons-material/{{member}}",
    },
    "@mui/lab": {
      transform: "@mui/lab/{{member}}",
    },

The modular imports knocked off 8 seconds alone

If you stumbled across this like i did for SST and AWS try also adding these settings to your SST config:
1) Memory setting of 3008 MB
2) Added setting warm: 20

**Reply by coolbii** [reply]
hi @bensleym 
do you mean all the imports in 'use client' components have to be replaced by dynamic import?
